### PR TITLE
[OTE-141] implement post /compliance/geoblock

### PR DESCRIPTION
**Pull request by dydxwill**
### Changelist
implement POST /compliance/geoblock logic

### Test Plan
unit tested.

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

### CODE REVIEW

File: indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts [hunk 1]
```diff
@@ -209,9 +211,89 @@ router.post(
         );
       }
 
-      // TODO(OTE-141): Implement logic.
+      /**
+       * If the address doesn't exist in the compliance table:
+       * - if the request is from a restricted country:
+       *  - if the action is ONBOARD, set the status to BLOCKED
+       *  - if the action is CONNECT, set the status to FIRST_STRIKE
+       * - else if the request is not from a non-restricted country:
```
Comment (vincentwschau) [inline indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts:219]:
```suggestion
       * - else if the request is from a non-restricted country:
```

File: indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts [hunk 2]
```diff
@@ -209,9 +211,89 @@ router.post(
...
+       *  - set the status to COMPLIANT
+       *
+       * if the address is COMPLIANT:
+       * - the ONLY action should be CONNECT
```
Comment (vincentwschau) [inline indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts:223]: nit: Note that if action is `ONBOARD` it is a no-op.

File: indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts [hunk 3]
```diff
@@ -322,5 +332,166 @@ describe('ComplianceV2Controller', () => {
...
+        type: RequestMethod.POST,
+        path: '/v4/compliance/geoblock',
+        body: {
+          address: testConstants.defaultAddress,
+          message: 'Test message',
+          action: ComplianceAction.CONNECT,
+          signedMessage: sha256(Buffer.from('msg')),
+          pubkey: new Uint8Array([/* public key bytes */]),
+          timestamp: 1620000000, // Valid timestamp
+        },
```
Comment (vincentwschau) [inline indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts:382]: nit: Pull out this body into a separate object so we don't have to repeat this per test-case?

File: indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts [hunk 4]
```diff
@@ -231,7 +236,12 @@ describe('ComplianceV2Controller', () => {
   });
 
   describe('POST /geoblock', () => {
```
Comment (vincentwschau) [inline indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts:238]: nit: Add test case of unexpected `ONBOARD` action, leading to error-log + no-op?

File: indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts (hunk 1 above)
Comment (dydxwill) [reply indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts:219]: done

File: indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts (hunk 2 above)
Comment (dydxwill) [reply indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts:223]: done

File: indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts [hunk 5]
```diff
@@ -209,9 +211,89 @@ router.post(
...
+              complianceStatus: complianceStatus[0],
+            });
+          } else if (
+            isRestrictedCountryHeaders(req.headers as CountryHeaders) &&
+            action === ComplianceAction.CONNECT
+          ) {
+            const complianceProgression: Partial<Record<ComplianceStatus, ComplianceStatus>> = {
+              [ComplianceStatus.COMPLIANT]: ComplianceStatus.FIRST_STRIKE,
+              [ComplianceStatus.FIRST_STRIKE]: ComplianceStatus.CLOSE_ONLY,
+            };
```
[the remark this answers is not in the document]
Comment (dydxwill) [reply indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts:283]: done

File: indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts (hunk 3 above)
Comment (dydxwill) [reply indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts:382]: done

File: indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts (hunk 4 above)
Comment (dydxwill) [reply indexer/services/comlink/__tests__/controllers/api/v4/compliance-v2-controller.test.ts:238]: done

File: indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts [hunk 6]
```diff
@@ -34,6 +36,11 @@ export enum ComplianceAction {
   CONNECT = 'CONNECT',
 }
 
+const complianceProgression: Partial<Record<ComplianceStatus, ComplianceStatus>> = {
```
Comment (vincentwschau) [inline indexer/services/comlink/src/controllers/api/v4/compliance-v2-controller.ts:39]: nit: As it's a constant config.
```suggestion
const COMPLIANCE_PROGRESSION: Partial<Record<ComplianceStatus, ComplianceStatus>> = {
```

**Review by vincentwschau — APPROVED** [review]

**Pull request merged**